File: src/ow/stdlib.pas
(* stdlib.pas -- Pascal declarations for stdlib.c.

   This file is for linking compiled object files with Pascal linker.
   It will be deleted in future when we rewrite the project in C.

   This is free and unencumbered software released into the public domain.
   For more information, please refer to <http://unlicense.org>. *)

unit stdlib;

interface

(*$I defines.pas*)

procedure malloc_;
procedure free_;

implementation

uses
    pascal;

(*$l stdlib.obj*)

procedure malloc_; external;
procedure free_; external;

end.

File: ow/stdlib.c
/* stdlib.c -- custom standard library.

   This is free and unencumbered software released into the public domain.
   For more information, please refer to <http://unlicense.org>. */

#ifdef __WATCOMC__
#include <i86.h>
#include <stdbool.h>
#include <stdint.h>
#endif

#include "../pascal/pascal.h"
#include "stdlib.h"

void *malloc(uint16_t size)
{
    uint16_t n;
    void *data;
    n = size + sizeof(uint16_t);
    pascal_getmem(&data, n);
    if (data)
    {
        *((uint16_t *)data) = n;
        data = MK_FP(FP_SEG(data), FP_OFF(data) + sizeof(uint16_t));
    };
    return data;
}

void free(void *ptr)
{
    void *data;
    data = MK_FP(FP_SEG(ptr), FP_OFF(ptr) - sizeof(uint16_t));
    pascal_freemem(data, *((uint16_t *)data));
}
